File: Packages/StackEngineCore/Tests/StackEngineCoreTests/ImageGeometryTests.swift
import XCTest
import simd
@testable import StackEngineCore

final class ImageGeometryTests: XCTestCase {
    // 3×2 image, each pixel uniquely valued by its row-major index.
    private func sample(_ w: Int = 3, _ h: Int = 2) -> PixelImage {
        PixelImage(width: w, height: h,
                   pixels: (0..<(w*h)).map { SIMD3(Float($0), Float($0) * 2, Float($0) * 3) })
    }

    func testZeroAndFourTurnsAreIdentity() {
        let img = sample()
        XCTAssertEqual(ImageGeometry.rotated(img, quarterTurns: 0), img)
        XCTAssertEqual(ImageGeometry.rotated(img, quarterTurns: 4), img)
        XCTAssertEqual(ImageGeometry.rotated(img, quarterTurns: -4), img)
    }

    func testOneTurnSwapsDimensionsAndMapsTopLeftToTopRight() {
        let img = sample(3, 2)                                  // w=3, h=2
        let r = ImageGeometry.rotated(img, quarterTurns: 1)     // 90° clockwise
        XCTAssertEqual(r.width, 2)
        XCTAssertEqual(r.height, 3)
        XCTAssertEqual(r[r.width - 1, 0], img[0, 0])           // source top-left → rotated top-right
    }

    func testTwoTurnsIs180() {
        let img = sample(3, 2)
        let r = ImageGeometry.rotated(img, quarterTurns: 2)
        XCTAssertEqual(r.width, 3); XCTAssertEqual(r.height, 2)
        XCTAssertEqual(r[img.width - 1, img.height - 1], img[0, 0])
    }

    func testNegativeWrapsToThree() {
        let img = sample()
        XCTAssertEqual(ImageGeometry.rotated(img, quarterTurns: -1),
                       ImageGeometry.rotated(img, quarterTurns: 3))
    }

    func testThreeTurnsSwapsDimensionsAndMapsTopLeftToBottomLeft() {
        let img = sample(3, 2)                                  // w=3, h=2
        let r = ImageGeometry.rotated(img, quarterTurns: 3)     // 270° CW = 90° CCW
        XCTAssertEqual(r.width, 2)
        XCTAssertEqual(r.height, 3)
        XCTAssertEqual(r[0, r.height - 1], img[0, 0])          // source top-left → rotated bottom-left
    }
}

File: android/stackengine/src/test/kotlin/com/jdpneto/stackengine/ImageGeometryTests.kt
package com.jdpneto.stackengine

import kotlin.test.Test
import kotlin.test.assertEquals

class ImageGeometryTests {

    // 3×2 image, each pixel uniquely valued by its row-major index.
    private fun sample(w: Int = 3, h: Int = 2): PixelImage {
        val px = FloatArray(w * h * 3)
        for (i in 0 until w * h) {
            px[i * 3] = i.toFloat(); px[i * 3 + 1] = i.toFloat() * 2f; px[i * 3 + 2] = i.toFloat() * 3f
        }
        return PixelImage(w, h, px)
    }

    @Test
    fun testZeroAndFourTurnsAreIdentity() {
        val img = sample()
        assertEquals(img, ImageGeometry.rotated(img, quarterTurns = 0))
        assertEquals(img, ImageGeometry.rotated(img, quarterTurns = 4))
        assertEquals(img, ImageGeometry.rotated(img, quarterTurns = -4))
    }

    @Test
    fun testOneTurnSwapsDimensionsAndMapsTopLeftToTopRight() {
        val img = sample(3, 2)                                       // w=3, h=2
        val r = ImageGeometry.rotated(img, quarterTurns = 1)         // 90° clockwise
        assertEquals(2, r.width)
        assertEquals(3, r.height)
        assertEquals(img[0, 0], r[r.width - 1, 0])                   // source top-left → rotated top-right
    }

    @Test
    fun testTwoTurnsIs180() {
        val img = sample(3, 2)
        val r = ImageGeometry.rotated(img, quarterTurns = 2)
        assertEquals(3, r.width); assertEquals(2, r.height)
        assertEquals(img[0, 0], r[img.width - 1, img.height - 1])
    }

    @Test
    fun testNegativeWrapsToThree() {
        val img = sample()
        assertEquals(ImageGeometry.rotated(img, quarterTurns = 3),
                     ImageGeometry.rotated(img, quarterTurns = -1))
    }

    @Test
    fun testThreeTurnsSwapsDimensionsAndMapsTopLeftToBottomLeft() {
        val img = sample(3, 2)                                       // w=3, h=2
        val r = ImageGeometry.rotated(img, quarterTurns = 3)         // 270° CW = 90° CCW
        assertEquals(2, r.width)
        assertEquals(3, r.height)
        assertEquals(img[0, 0], r[0, r.height - 1])                  // source top-left → rotated bottom-left
    }
}
